search: create a search result dedupper and update unionMerge

camdencheek opened the pull request:
Stacked on: #18191 
Reopened from the reverted #18148 

This creates the `searchResultDedupper` type, which facilitates
deduplicating search results by URL. It replaces the more complex logic
in `unionMerge` and will also be usable for deduplicating select
results.


<!-- Reminder: Have you updated the changelog and relevant docs (user docs, architecture diagram, etc) ? -->

camdencheek commented on cmd/frontend/graphqlbackend/search_results.go:
```diff
@@ -520,92 +520,87 @@ func (r *searchResolver) evaluateLeaf(ctx context.Context) (_ *SearchResultsReso
+		rightFileMatch, ok1 := rightResult.ToFileMatch()
+		leftFileMatch, ok2 := prev.ToFileMatch()
+		if ok1 && ok2 {
+			leftFileMatch.appendMatches(rightFileMatch)
```
This fixes the cause for the previous revert. There are cases in `TestUnionMerge` that specifically hit this.

camdencheek replied:
Side note: is there a reason `appendMatches` doesn't deduplicate the line matches or symbol matches? Seems like this would cause duplicate results.

rvantonder replied:
Heh, looks like you're right about that! But note, you should turn off streaming to see what is truly being duplicated so it's not conflated with what's happening in #18203. When streaming is off, it's still wrong, but it doesn't look as wrong, there is only one visible match:

![Screen Shot 2021-02-12 at 12 17 47 AM](https://user-images.githubusercontent.com/888624/107740188-c1c2e880-6cc7-11eb-82e8-a0381fd831a1.png)

However, you can tell it's doing something wrong because of the match count (2).

A more obviously wrong example happens with:

[`repo:^github\.com/sourcegraph/sourcegraph$ search or search or search or search file:parserFuzz.ts count:1000`](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+search+or+search+or+search+or+search+file:parserFuzz.ts+count:1000&patternType=literal).

I don't feel terrible about this last example, because it's a bit silly to search for the same thing. The regex example is more of a real case, but it does prompt whether overlapping matches for single terms should "logically" be counted twice, or not. If you look at the regex equivalent, we just throw away matches that would match the alternative, so it doesn't exactly honor the whole expression (it just doesn't try to deal with this problem):

[`repo:^github\.com/sourcegraph/sourcegraph$ (search|search\ query) file:parserFuzz.ts count:1000`](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+%28search%7Csearch%5C+query%29+file:parserFuzz.ts+count:1000&patternType=regexp)

Out of curiosity I looked at what `cs.android.com` does because they support `or` in this sense. It seems to take the longest contiguous range as one match (the match `int test passed` is counted once, even though there is no search term for `int test passed`, only `int test` and `test_passed` separately):

https://cs.android.com/search?q=test%20OR%20test_passed%20OR%20%22int%20test%22%20file:test.h

<img width="1250" alt="Screen Shot 2021-02-12 at 12 28 50 AM" src="https://user-images.githubusercontent.com/888624/107740869-4c581780-6cc9-11eb-8b7a-d02a6e13b12a.png">

My take: it's worthwhile deduping line matches, but only if they are non-overlapping. If we consider partial range overlaps, it gets icky to answer how to dedupe (my intuition is just to include both ranges, if there is partial overlap--the alternative is to do a merge, like cs.android.com, but I don't think we strictly need to follow suit here).

OK this got long and deserves its own issue. I'll create one tomorrow. I think we want to just have an alternative to `appendMatches` that dedupes easy things at the line level (equal ranges, or if one range completely encloses another), and includes both if there's a partial overlap.

rvantonder reviewed (commented):
I reviewed parts of this and need more time to look over things.  I'm currently a bit iffy on `dedup.Seen` because it's return value is sometimes ignored, and sometimes is not, so it's not side-effect free. The dual purpose of the single `Seen` function (sending a signal about whether something is seen, and internal tracking to update state) is a bit confusing. I'm currently bouncing between this and #18073, because I see what it is setting up to do in #18073, while loop over results, but it feels like there should be a better way--will come back to this tomorrow.

camdencheek commented:
>  I'm currently a bit iffy on dedup.Seen because it's return value is sometimes ignored, and sometimes is not, so it's not side-effect free

Yeah, I totally see what you mean. We've got mutations happening across logical ownership boundaries. I think I'll try moving all the dedup logic into the dedupper.

camdencheek commented:
@rvantonder In response to your comments, I refactored `dedupper` a bit more (and renamed it to `deduper`). I think the new interface is considerably simpler and easier to understand. I think it also resolves all the ownership problems

rvantonder commented on cmd/frontend/graphqlbackend/search_results.go:
```diff
@@ -520,95 +520,98 @@ func (r *searchResolver) evaluateLeaf(ctx context.Context) (_ *SearchResultsReso
+	return
+}
+
+// Deduped returns a slice of SearchResultResolvers, deduplicated from
+// the SearchResultResolvers that were added with Add
+func (d *searchResultDeduper) Deduped() []SearchResultResolver {
```
```suggestion
// Dedupe deduplicates SearchResultResolvers that were added with Add.
func (d *searchResultDeduper) Dedupe() []SearchResultResolver {
```

Nit: would just make this a verb.

rvantonder commented on cmd/frontend/graphqlbackend/search_results.go:
```diff
@@ -520,95 +520,98 @@ func (r *searchResolver) evaluateLeaf(ctx context.Context) (_ *SearchResultsReso
+func unionMerge(left, right *SearchResultsResolver) *SearchResultsResolver {
+	var dedup searchResultDeduper
+
+	// Register all the left results in the dedupper
+	for _, leftResult := range left.SearchResults {
+		dedup.Add(leftResult)
+	}
+
+	for _, rightResult := range right.SearchResults {
+		dedup.Add(rightResult)
+	}
```
```suggestion
	// Add results to maps for deduping.
	for _, leftResult := range left.SearchResults {
		dedup.Add(leftResult)
	}
	for _, rightResult := range right.SearchResults {
		dedup.Add(rightResult)
	}
```

Nit: comment seems unbalanced talking about left and not right :-)

rvantonder commented on cmd/frontend/graphqlbackend/search_results.go:
```diff
@@ -520,95 +520,98 @@ func (r *searchResolver) evaluateLeaf(ctx context.Context) (_ *SearchResultsReso
-			continue
+			d.seenDiffMatches[commitMatch.URL()] = commitMatch
+			return
+		} else {
+			if _, seen := d.seenCommitMatches[commitMatch.URL()]; seen {
+				return
+			}
+			d.seenCommitMatches[commitMatch.URL()] = commitMatch
+			return
```
```suggestion
		}
		if _, seen := d.seenCommitMatches[commitMatch.URL()]; seen {
			return
		}
		d.seenCommitMatches[commitMatch.URL()] = commitMatch
		return
```

Nit: Pretty sure we don't need this in an `else` branch.

rvantonder commented on cmd/frontend/graphqlbackend/search_results.go:
```diff
@@ -520,95 +520,98 @@ func (r *searchResolver) evaluateLeaf(ctx context.Context) (_ *SearchResultsReso
 
-	for _, v := range rightFileMatches {
-		merged = append(merged, v)
+	return
```
What lint says
```suggestion
```

rvantonder commented on go.sum:
```diff
@@ -308,10 +308,13 @@ github.com/go-enry/go-enry/v2 v2.6.0 h1:nbGWQBpO+D+cJuRxNgSDFnFY9QWz3QM/CeZxU7VA
 github.com/go-enry/go-enry/v2 v2.6.0/go.mod h1:GVzIiAytiS5uT/QiuakK7TF1u4xDab87Y8V5EJRpsIQ=
 github.com/go-enry/go-oniguruma v1.2.1/go.mod h1:bWDhYP+S6xZQgiRL7wlTScFYBe023B6ilRZbCAD5Hf4=
 github.com/go-errors/errors v1.0.1/go.mod h1:f4zRHt4oKfwPJE5k8C9vpYG+aDHdBFUsgrm6/TyX73Q=
+github.com/go-git/gcfg v1.5.0 h1:Q5ViNfGF8zFgyJWPqYwA7qGFoMTEiBmdlkcfRmpIMa4=
```
If you rebase on `main` these `go.sum` changes should disappear.

rvantonder commented on cmd/frontend/graphqlbackend/search_results.go:
```diff
@@ -520,95 +520,98 @@ func (r *searchResolver) evaluateLeaf(ctx context.Context) (_ *SearchResultsReso
+// Add adds a SearchResultResolver to the deduper, merging it into
+// a previously added SearchResultResolver if the URL has already been seen
+func (d *searchResultDeduper) Add(r SearchResultResolver) {
+	d.init()
```
This is a nit, but I think it's cleaner to just create a `NewDeduper` function, since it looks like we really only need to call init once at the start of when we do deduping, local to the functions like `unionMerge` and `selectResult`.

Then we just change the `var dedup searchResultDeduper` to be `deduper := NewDeduper()` in those functions. Using `NewThing` is, as far as I know, fairly idiomatic in `Go`, we use it, e.g., for `NewSearchImplementer`.

camdencheek replied:
Ha, yeah. I would normally do that, but I was experimenting with some different patterns. I agree that in this case, `New...` is simpler and more idiomatic. I'll make the swap

rvantonder reviewed (approved):
I like the `Deduper` muuuuch better now 😍 . Only nits--high level thought: what do you think about putting `deduper` in it's own file like `result_deduper.go`? Just feel like `search_results.go` is super bloated.

camdencheek commented:
@rvantonder Nits addressed, totally agree that it makes sense in its own file
